src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+                )
+                wfm = i_env * sinalpha + q_env * cosalpha
+
+                # First we convert the pulse starting time to number of ADC samples
+                # Then, we convert this number to the number of ADC clock cycles (8 samples per clock cycle)
+                # Next, we raise it to the next nearest integer to prevent an overlap between drive and readout pulses
+                # Finally, we ensure that the number is even for the DAC delay conversion
+                delay_start_adc = int(
+                    int(
+                        np.ceil(
+                            self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9 / 8
+                        )
+                        / 2
+                    )
+                    * 2
+                )
 
-    def stop(self):
-        self._send(b"stop")
+                # For the DAC, currently the sampling rate is 3x higher than the ADC
+                # The number of clock cycles is 16 samples per clock cycle
+                # Hence, we multiply the adc delay clock cycles by 1.5x to align the DAC/ADC pair
+                delay_start_dac = int(delay_start_adc * 1.5)
 
-    def status(self):
-        return self._send(b"status", True)
+                self.device.dac[dac].delay = (
+                    delay_start_dac + self.channel_delay_offset_dac
+                )
+                self.device.adc[adc].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
+                # ADC0 complete marks the end of acquisition, so we also need to move ADC0
+                self.device.adc[0].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
 
-    def _send(self, payload, retval=False):
-        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-            s.connect((self.address, self.port))
-            s.sendall(payload)
+                if (
+                    options.acquisition_type is AcquisitionType.DISCRIMINATION
+                    or AcquisitionType.INTEGRATION
+                ):
+                    self.device.program_qunit(
+                        readout_frequency=pulse.frequency,
+                        readout_time=pulse.duration * 1e-9,
+                        qunit=pulse.qubit,
+                    )
+
+            end = start + len(wfm)
+            waveform_array[dac][start:end] += self.device.dac_max_amplitude * wfm
+            dac_end_addr[dac] = max(end >> 4, dac_end_addr[dac])
+
+        payload = [
+            (dac, wfm, dac_end_addr[dac])
+            for dac, wfm in waveform_array.items()
+            if dac_end_addr[dac] != 0
+        ]
+        self.device.upload_waveform(payload)
+
+    def play_sequences(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+    ):
+        pass
```

Comment:
This is not part of the interface, and never used

https://github.com/search?q=repo%3Aqiboteam%2Fqibolab%20play_sequences&type=code

So, I'd drop it.
```suggestion
```

(at platform level `execute_pulse_sequences` exists, but it is implemented through the unrolling, so it doesn't require a matching method at `Controller` level, only batching support)

Reply:
Resolved in c08af5d